File: frontend/src/types/trip.ts
// Trip-related types based on backend DTOs
export interface Translations {
  th?: {
    title?: string
    description?: string
  }
  en?: {
    title?: string
    description?: string
  }
}

export interface Trip {
  id: number
  title: string
  description?: string
  translations?: Translations
  photos: string[]
  tags: string[]
  latitude?: number | null
  longitude?: number | null
  authorId?: number
  authorName?: string
  createdAt?: string
  updatedAt?: string
  province?: string
}

export interface TripSummary {
  id: number
  title: string
  shortDescription?: string
  translations?: Translations
  coverImage?: string
  province?: string
  tags: string[]
  authorId?: number
  updatedAt?: string
}

export interface CreateTripData {
  title: string
  description?: string
  photos: string[]
  tags: string[]
  latitude?: number | null
  longitude?: number | null
}

export interface FormData {
  title: string
  description: string
  photos: string[]
  tags: string[]
  latitude: number | null
  longitude: number | null
}

File: backend/src/main/java/com/travelapp/entity/Trip.java
package com.travelapp.entity;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.Data;

@Entity
@Table(name = "trips")
@Data
public class Trip {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String title; // Keep for backward compatibility, will use translations if available

    @Column(columnDefinition = "TEXT")
    private String description; // Keep for backward compatibility, will use translations if available

    // Translations column - stores multilingual content as JSONB
    @Column(name = "translations", columnDefinition = "JSONB")
    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, Map<String, String>> translations;

    @Column(name = "photos", columnDefinition = "TEXT[]")
    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> photos = new ArrayList<>();

    @Column(name = "tags", columnDefinition = "TEXT[]")
    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> tags = new ArrayList<>();

    @Column(name = "latitude")
    private Double latitude;

    @Column(name = "longitude")
    private Double longitude;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", foreignKey = @ForeignKey(name = "fk_trips_author"))
    private User author;

    @Column(name = "author_id", insertable = false, updatable = false)
    private Long authorId;

    @Column(name = "created_at", updatable = false, columnDefinition = "TIMESTAMPTZ")
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", columnDefinition = "TIMESTAMPTZ")
    private OffsetDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = OffsetDateTime.now();
        }
        updatedAt = OffsetDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}

